### ISSUE TITLE
Change how merges are done

### ISSUE THREAD

**Issue by spacek531** [OpenRCT2/OpenRCT2#15397]
> I feel that [merges] be supported natively as well through some sort of flag. The get next track piece code relies on just so much bad logic at the moment and it would be nice if it could be simplified and made safe. Will be a pain to work out how to convert saves though.

The question is how should such a flag behave, and once the flag behavior is determined, what needs to be done to import saves?

**Comment by spacek531** [comment]
I have pondered the feasibility of pointing at another specific TileElement. If it's feasible, then it's possible knock out two birds with one stone by teleporting the car from the given TrackElement onto the selected TrackElement. For merging, the user would select an adjacent track piece the vehicle should travel to, including another ride's. For teleporting, the user would select a non-adjacent track piece the vehicle should travel to.

If this method is selected, being able to choose a forward and backward direction would make this particularly powerful.

**Comment by duncanspumpkin** [comment]
You are describing a doubly linked list. Yes that would be a possible solution but it would not be possible to encode this into the track element. You could do it as some additional structure tied to the ride though. It would get messy though trying to work out references.

**Comment by spacek531** [comment]
I don't think this is a doubly linked list because it would be used only for merges.

**Comment by duncanspumpkin** [comment]
In general you should avoid using pointers except when you absolutely have to. Serialisation, invalidation, memory safety are all major issues when you start using pointers.

**Comment by spacek531** [comment]
If the discussion on handles https://github.com/OpenRCT2/OpenRCT2/issues/15417 leads somewhere, changing merges to a feature that requires user intent is possible.

A merges field (or just teleport field) for forward and backward directions would live in the extended track data. As discussed in the linked PR, when the train exits the track element going that direction and calls TrackMotionGetNewTrack, it immediately shortcuts to the linked track. The complicated 'also have collision detection' maybe doesn't even have to happen.

If SV6 import or NSF loading is done on the datastream, an extra step has to be added to touch all the TileElements of older save versions after the data is loadedf. I think SV6 streams the data before import, which means populating the merge/teleport field of extended track data can be done using SV6 data and not create an extra step after importing TileElement data.

**Comment by spacek531** [comment]
There's another bug I found when pouring over the source code.
When the down25toflat piece is built before the up25 piece, the coaster crashes
![image](https://user-images.githubusercontent.com/3454156/133919870-cd5b142b-769e-4eec-9a2a-b4f4837e0a93.png)

`track_block_get_next_from_zero` and `track_block_get_previous_from_zero` do not take into account the bank nor slope of the pieces.
